xymond/hobbitd_capture.8
.TH HOBBITD_CAPTURE 8 "Version 4.2.0:  9 Aug 2006" "Hobbit Monitor"
.SH NAME
hobbitd_capture \- catch selected messages from a hobbitd channel
.SH SYNOPSIS
.B "hobbitd_channel --channel=status hobbitd_capture [options]"

.SH DESCRIPTION
hobbitd_capture is a worker module for hobbitd, and as such it is normally
run via the
.I hobbitd_channel(8)
program. It receives messages from hobbitd via stdin and filters them to
select messages based on the hostname, testname or color of the status.
By default the resulting messages are printed on stdout, but they can also
be fed into a command for further processing.

hobbitd_capture supports the \fBstatus\fR, \fBdata\fR and \fBhostdata\fR
channels.

.SH OPTIONS
.IP "--hosts=PATTERN"
Select messages only from hosts matching PATTERN (regular expression).

.IP "--exhosts=PATTERN"
Exclude messages from hosts matching PATTERN. If used with the --hosts option, 
then the hostname must match the --hosts pattern, but NOT the --exhosts pattern.

.IP "--tests=PATTERN"
Select messages only from tests matching PATTERN (regular expression).

.IP "--extests=PATTERN"
Exclude messages from tests matching PATTERN. If used with the --tests option, 
then the testname must match the --tests pattern, but NOT the --extests pattern.

.IP "--colors=COLOR[,color]"
Select messages based on the color of the status message. Multiple colors can be 
listed, separated by comma. Default: Accept all colors.

.IP "--batch-command=COMMAND"
Instead of printing the messages to stdout, feed them to COMMAND on stdin. COMMAND
can be any command which accepts the mssage on standard input.

.IP "--batch-timeout=SECONDS"
Collect messages until no messages have arrived in SECONDS seconds, before sending
them to the --batch-command COMMAND.

.IP "--debug"
Enable debugging output.

.SH "SEE ALSO"
hobbitd_channel(8), hobbitd(8), hobbit(7)



Changes
Changes from 4.3.0 RC 1 -> 4.3.0 (4 Mar 2011)
=============================================
* Critical view and other webpages: Make the 'All systems OK' message 
  configurable. Also allow the header/footer for the Critical Systems 
  view to be configurable.
  Suggestion and preliminary patch from Buchan Milne.
* xymonnet: Improve error report when HTTP tests get an empty response - 
  'HTTP error 0' sounds weird.
* report / snapshot CGI's: Fix buffer overrun in the HTML delimiter 
  generated in the "please wait..." message. Also, fix potential buffer 
  overrun in report CGI if invoked with a large value for the "style" 
  parameter. 
  Reported by Rolf Biesbroek.
* Graph definitions (graphs.cfg): Multi graphs cannot use a regex pattern. 
  Problem report by Brian Majeska
* Solaris interface statistics: Filter out "mac" and "wrsmd" devices at
  the client side. 
  Update RRD handler to also filter "wrsmd" at the server side, like we 
  already did for "mac" devices. 
  Cf. http://www.xymon.com/archive/2009/06/msg00204.html
* Documentation: Document the XMH_* fields available in xymondboard commands.
* Documentation: Document SPLITNCV and "trends" methods of doing 
  custom graphs.
* RRD definitions: Allow override of --step/-s option for rrdcreate, 
  from template supplied in rrddefinitions.cfg. 
  Suggestion from Brian Majeska.
* mailack: Remove restriction on how long a subjectline/message body can be.
* Build procedure: Add notice about running upgrade script before 
  installing the new version.
* xymond_alert: Document --trace option
* Alerts: For recovery messages, add information so you can tell 
  whether the recovery was due to the service actually recovering, or 
  if it was merely disabled.
* xymond_alert: Fix missing element in array of alert status texts used 
  for tracing. 
  Spotted by Dominique Frise.
* Add support for FreeBSD v8 modified ifstat output
* Documentation: Update information about the Xymon mailing lists 
  following move to Mailman and new archive URL.
* HP/UX client: Use "swapinfo" to extract memory utilisation data, 
  instead of the hpux-meminfo utility. 
  By Earl Flack http://lists.xymon.com/pipermail/xymon/2010-December/030100.html

Changes from 4.3.0 beta 3 -> 4.3.0 RC 1 (23 Jan 2011)
=====================================================
* hosts.cfg badldap documentation: Document that for LDAP URL's you must 
  use 'badldapurl'. Reported by Simo Hmami.
* xymond flap detection: Make number of tracked status changes and the 
  flap-check period configurable. Change the defaults to trigger flapping 
  at more than 5 status changes in a 30 minute period.
* sendmessage: Enhanced error reporting, to help track down communication 
  problems.
* xymond_client: Fix Windows SVC status handling to avoid coredumps, 
  memory corruption and other nasties. 
  Will now report the real name of the service, instead of the pattern used in 
  the analysis.cfg file.
  NOTE: Slight change to status message format.
* Client handler: Fix owner/user check parsing. Reported by Ian Marsh 
  http://www.xymon.com/archive/2011/01/msg00133.html (also broken in 4.2.3).
* xymongen: Fix broken --doc-window option handling. Reported by Tom Schmitt.
* Xymongen: Fix documentation of the --doc-window/--no-doc-window options.
* Webpage background: Use a CSS and a new set of gif's to implement a 
  background that works on all displays, regardless of width. Uses a new 
  xymonbody.css stylesheet which can also control some other aspects of the 
  webpage. From Francois Claire.
* Documentation: The xymon 'rename' command should be used AFTER renaming 
  a host in hosts.cfg, not before. From Tom Georgoulias.
* Memory status: Add some sanity checks for the memory utilisation reported 
  by clients. Occasionally we get completely bogus data from clients, so only 
  act on them if percentages do not exceed 100.
* Critical systems view: Add "--tooltips" option so you can save screen space 
  by hiding the host descriptions in a tooltip, like we do on the statically 
  generated pages. Feature request from Chris Morris.
* Solaris client: Report "swap -l" in addition to "swap -s" for swap usage. 
  Backend prefers output from "swap -l" when determining swap utilisation.
* Webpage menu: Use the CSS and GIF's by Malcolm Hunter - they are much nicer 
  than the ones from Debian. Distribute both the blue and the grey version, 
  and configure which one to use in xymonserver.cfg.
* Graph zoom: Use float variables when calculating the upper/lower limits 
  of the graph. Fixes vertical zoom.
* xymond: Make sure we do not perform socket operations on invalid sockets 
  (e.g. those from a scheduled task pseudo-connection)
* Installation: Remove any existing old commands before creating symlinks
* xymonproxy: Fix broken compatibility option '--bbdisplay'
* Fix eventlog summary/count enums so they dont clash with Solaris predefined 
  entities
* History- and hostdata-modules: Dont save data if there is less than 5% 
  free space on the filesystem. Also, dont save hostdata info more than 5 
  times per hour.
* Historical statuslog display: Work-around for crash when status-log is 
  empty
* fping.sh configure sub-script: Fix syntax error in suggested 'sudoers' 
  configuration, and point to the found fping binary. From Steff Coene.
* namematch routine: Fix broken matching when doing simple matching against 
  two strings where one was a subset of the other. 
  http://www.xymon.com/archive/2010/11/msg00177.html .  Reported by Elmar Heeb 
  who also provided a patch, although I chose a different solution to this.
* Xymon net: Fix broken compile when LDAP-checks are disabled. Reported by 
  Roland Soderstrom, fix from Ralph Mitchell.
* xymon(7) manpage: Drop notice that renaming in 4.3.0 is not complete
* Installation: Setup links for the commonly used Hobbit binaries 
  (bb, bbcmd, bbdigest, bbhostgrep, bbhostshow)
* Upgrade script: Setup symlinks for the old names of the standard webpages

xymond/hobbitd_capture.c
/*----------------------------------------------------------------------------*/
/* Hobbit message daemon.                                                     */
/*                                                                            */
/* Hobbitd worker module, to capture status messages for a particulr host     */
/* or test, or data type. This is fed from the status- or data-channel, and   */
/* simply logs the data received to a file.                                   */
/*                                                                            */
/* Copyright (C) 2004-2008 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/time.h>
#include <errno.h>

#include "libbbgen.h"
#include "hobbitd_worker.h"

#include <signal.h>


#define MAX_META 20	/* The maximum number of meta-data items in a message */


int main(int argc, char *argv[])
{
	char *msg;
	int running;
	int argi, seq;
	struct timeval *timeout = NULL;
	pcre *hostexp = NULL;
	pcre *exhostexp = NULL;
	pcre *testexp = NULL;
	pcre *extestexp = NULL;
	pcre *colorexp = NULL;
        const char *errmsg = NULL;
	int errofs = 0;
	FILE *logfd = stdout;
	int batchtimeout = 30;
	char *batchcmd = NULL;
	strbuffer_t *batchbuf = NULL;
	time_t lastmsgtime = 0;

	/* Handle program options. */
	for (argi = 1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--debug") == 0) {
			/*
			 * A global "debug" variable is available. If
			 * it is set, then "dbgprintf()" outputs debug messages.
			 */
			debug = 1;
		}
		else if (strncmp(argv[argi], "--timeout=", 10) == 0) {
			/*
			 * You can have a timeout when waiting for new
			 * messages. If it happens, you will get a "@@idle\n"
			 * message with sequence number 0.
			 * If you dont want a timeout, just pass a NULL for the timeout parameter.
			 */
			timeout = (struct timeval *)(malloc(sizeof(struct timeval)));
			timeout->tv_sec = (atoi(argv[argi]+10));
			timeout->tv_usec = 0;
		}
		else if (argnmatch(argv[argi], "--hosts=")) {
			char *exp = strchr(argv[argi], '=') + 1;
			hostexp = pcre_compile(exp, PCRE_CASELESS, &errmsg, &errofs, NULL);
			if (hostexp == NULL) printf("Invalid expression '%s'\n", exp);
		}
		else if (argnmatch(argv[argi], "--exhosts=")) {
			char *exp = strchr(argv[argi], '=') + 1;
			exhostexp = pcre_compile(exp, PCRE_CASELESS, &errmsg, &errofs, NULL);
			if (exhostexp == NULL) printf("Invalid expression '%s'\n", exp);
		}
		else if (argnmatch(argv[argi], "--tests=")) {
			char *exp = strchr(argv[argi], '=') + 1;
			testexp = pcre_compile(exp, PCRE_CASELESS, &errmsg, &errofs, NULL);
			if (testexp == NULL) printf("Invalid expression '%s'\n", exp);
		}
		else if (argnmatch(argv[argi], "--extests=")) {
			char *exp = strchr(argv[argi], '=') + 1;
			extestexp = pcre_compile(exp, PCRE_CASELESS, &errmsg, &errofs, NULL);
			if (extestexp == NULL) printf("Invalid expression '%s'\n", exp);
		}
		else if (argnmatch(argv[argi], "--colors=")) {
			char *exp = strchr(argv[argi], '=') + 1;
			colorexp = pcre_compile(exp, PCRE_CASELESS, &errmsg, &errofs, NULL);
			if (colorexp == NULL) printf("Invalid expression '%s'\n", exp);
		}
		else if (argnmatch(argv[argi], "--outfile=")) {
			char *fn = strchr(argv[argi], '=') + 1;
			logfd = fopen(fn, "a");
			if (logfd == NULL) {
				printf("Cannot open logfile %s: %s\n", fn, strerror(errno));
				logfd = stdout;
			}
		}
		else if (argnmatch(argv[argi], "--batch-timeout=")) {
			char *p = strchr(argv[argi], '=');
			batchtimeout = atoi(p+1);
			timeout = (struct timeval *)(malloc(sizeof(struct timeval)));
			timeout->tv_sec = batchtimeout;
			timeout->tv_usec = 0;
		}
		else if (argnmatch(argv[argi], "--batch-command=")) {
			char *p = strchr(argv[argi], '=');
			batchcmd = strdup(p+1);
			batchbuf = newstrbuffer(0);
		}
		else {
			printf("Unknown option %s\n", argv[argi]);
			printf("Usage: %s [--hosts=EXP] [--tests=EXP] [--exhosts=EXP] [--extests=EXP] [--color=EXP] [--outfile=FILENAME] [--batch-timeout=N] [--batch-command=COMMAND]\n", argv[0]);
			return 0;
		}
	}

	signal(SIGCHLD, SIG_IGN);

	running = 1;
	while (running) {
		char *eoln, *restofmsg, *p;
		char *metadata[MAX_META+1];
		int metacount;

		msg = get_hobbitd_message(C_LAST, argv[0], &seq, timeout);
		if (msg == NULL) {
			/*
			 * get_hobbitd_message will return NULL if hobbitd_channel closes
			 * the input pipe. We should shutdown when that happens.
			 */
			running = 0;
			continue;
		}

		/*
		 * Now we have a message. So do something with it.
		 *
		 * The first line of the message is always a '|' separated
		 * list of meta-data about the message. After the first
		 * line, the content varies by channel.
		 */

		/* Split the message in the first line (with meta-data), and the rest */
 		eoln = strchr(msg, '\n');
		if (eoln) {
			*eoln = '\0';
			restofmsg = eoln+1;
		}
		else {
			restofmsg = "";
		}

		/* 
		 * Now parse the meta-data into elements.
		 * We use our own "gettok()" routine which works
		 * like strtok(), but can handle empty elements.
		 */
		metacount = 0; 
		p = gettok(msg, "|");
		while (p && (metacount < MAX_META)) {
			metadata[metacount++] = p;
			p = gettok(NULL, "|");
		}
		metadata[metacount] = NULL;

		/*
		 * A "shutdown" message is sent when the master daemon
		 * terminates. The child workers should shutdown also.
		 */
		if (strncmp(metadata[0], "@@shutdown", 10) == 0) {
			printf("Shutting down\n");
			running = 0;
			continue;
		}

		/*
		 * A "logrotate" message is sent when the Hobbit logs are
		 * rotated. The child workers must re-open their logfiles,
		 * typically stdin and stderr - the filename is always
		 * provided in the HOBBITCHANNEL_LOGFILENAME environment.
		 */
		else if (strncmp(metadata[0], "@@logrotate", 11) == 0) {
			char *fn = xgetenv("HOBBITCHANNEL_LOGFILENAME");
			if (fn && strlen(fn)) {
				freopen(fn, "a", stdout);
				freopen(fn, "a", stderr);
			}
			continue;
		}

		/*
		 * An "idle" message appears when get_hobbitd_message() 
		 * exceeds the timeout setting (ie. you passed a timeout
		 * value). This allows your worker module to perform
		 * some internal processing even though no messages arrive.
		 */
		else if (strncmp(metadata[0], "@@idle", 6) == 0) {
			dbgprintf("Got an 'idle' message\n");
		}

		/*
		 * The "drophost", "droptest", "renamehost" and "renametst"
		 * indicates that a host/test was deleted or renamed. If the
		 * worker module maintains some internal storage (in memory
		 * or persistent file-storage), it should act on these
		 * messages to maintain data consistency.
		 */
		else if ((metacount > 3) && (strncmp(metadata[0], "@@drophost", 10) == 0)) {
			dbgprintf("Got a 'drophost' message for host '%s'\n", metadata[3]);
		}
		else if ((metacount > 3) && (strncmp(metadata[0], "@@dropstate", 11) == 0)) {
			dbgprintf("Got a 'dropstate' message for host '%s'\n", metadata[3]);
		}
		else if ((metacount > 4) && (strncmp(metadata[0], "@@droptest", 10) == 0)) {
			dbgprintf("Got a 'droptest' message for host '%s' test '%s'\n", metadata[3], metadata[4]);
		}
		else if ((metacount > 4) && (strncmp(metadata[0], "@@renamehost", 12) == 0)) {
			dbgprintf("Got a 'renamehost' message for host '%s' -> '%s'\n", metadata[3], metadata[4]);
		}
		else if ((metacount > 5) && (strncmp(metadata[0], "@@renametest", 12) == 0)) {
			dbgprintf("Got a 'renametest' message for host '%s' test '%s' -> '%s'\n", 
				metadata[3], metadata[4], metadata[5]);
		}

		/*
		 * Process this message.
		 */
		else {
			int ovector[30];
			int match, i;
			char *hostname = metadata[4];
			char *testname = metadata[5];
			char *color = metadata[7];

			/* See if we should handle the batched messages we've got */
			if (batchcmd && ((lastmsgtime + batchtimeout) < getcurrenttime(NULL)) && (STRBUFLEN(batchbuf) > 0)) {
				pid_t childpid = fork();
				int childres = 0;

				if (childpid < 0) {
					/* Fork failed! */
					errprintf("Fork failed: %s\n", strerror(errno));
				}
				else if (childpid == 0) {
					/* Child */
					FILE *cmdpipe = popen(batchcmd, "w");
					if (cmdpipe) {
						/* Write the data to the batch command pipe */
						int n, bytesleft = STRBUFLEN(batchbuf);
						char *outp = STRBUF(batchbuf);

						while (bytesleft) {
							n = fwrite(outp, 1, bytesleft, cmdpipe);
							if (n >= 0) {
								bytesleft -= n;
								outp += n;
							}
							else {
								errprintf("Error while writing data to batch command\n");
								bytesleft = 0;
							}
						}

						childres = pclose(cmdpipe);
					}
					else {
						errprintf("Could not open pipe to batch command '%s'\n", batchcmd);
						childres = 127;
					}

					exit(childres);
				}
				else if (childpid > 0) {
					/* Parent continues */
				}

				clearstrbuffer(batchbuf);
			}


			if (hostexp) {
				match = (pcre_exec(hostexp, NULL, hostname, strlen(hostname), 0, 0, ovector, (sizeof(ovector)/sizeof(int))) >= 0);
				if (!match) continue;
			}
			if (exhostexp) {
				match = (pcre_exec(exhostexp, NULL, hostname, strlen(hostname), 0, 0, ovector, (sizeof(ovector)/sizeof(int))) >= 0);
				if (match) continue;
			}
			if (testexp) {
				match = (pcre_exec(testexp, NULL, testname, strlen(testname), 0, 0, ovector, (sizeof(ovector)/sizeof(int))) >= 0);
				if (!match) continue;
			}
			if (exhostexp) {
				match = (pcre_exec(extestexp, NULL, testname, strlen(testname), 0, 0, ovector, (sizeof(ovector)/sizeof(int))) >= 0);
				if (match) continue;
			}
			if (colorexp) {
				match = (pcre_exec(colorexp, NULL, color, strlen(color), 0, 0, ovector, (sizeof(ovector)/sizeof(int))) >= 0);
				if (!match) continue;
			}

			lastmsgtime = getcurrenttime(NULL);

			if (batchcmd) {
				addtobuffer(batchbuf, "## ");
				for (i=0; (i < metacount); i++) {
					addtobuffer(batchbuf, metadata[i]);
					addtobuffer(batchbuf, " ");
				}
				addtobuffer(batchbuf, "\n");
				addtobuffer(batchbuf, restofmsg);
				addtobuffer(batchbuf, "\n");
			}
			else {
				fprintf(logfd, "## ");
				for (i=0; (i < metacount); i++) fprintf(logfd, "%s ", metadata[i]);
				fprintf(logfd, "\n");
				fprintf(logfd, "%s\n", restofmsg);
			}
		}
	}

	return 0;
}

